### PR TITLE
feature: Introduce pre-commit for linting and formatting

### PR DESCRIPTION
**Pull request by harshad16**
feature: Introduce pre-commit for linting and formatting

## Description

Introducing pre-commit, which will bring in linting to the notebook repos.
Related-to: https://github.com/opendatahub-io/notebooks/issues/248

## How Has This Been Tested?

Bring in linting: 
1. Installation process: https://pre-commit.com/#install
2. https://pre-commit.com/#quick-start

Testing would require, testing the notebook image based on this.

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: ci/check-json.sh [hunk 1]
```diff
@@ -24,13 +24,13 @@ function check_json() {
...
         jsons=$(yq -r ".spec.tags[].annotations.\"${string}\"" "${f}")
-        
+
         while IFS= read -r json; do
             echo "    ${json}"
             echo -n "  > "; echo "${json}" | json_verify || ret_code="${?}"
         done <<< "${jsons}"
     else
-	echo "    Ignoring as this file doesn't contain necessary key field '${string}' for check"
+    echo "    Ignoring as this file doesn't contain necessary key field '${string}' for check"
```
Comment (jstourac) [inline ci/check-json.sh:33]: hm, this change is weird

File: jupyter/datascience/ubi9-python-3.9/utils/processor_kfp.patch [hunk 2]
```diff
@@ -9,30 +9,30 @@
...
- 
+
              except Exception as ex:
 @@ -435,7 +436,7 @@
- 
+
              self.log_pipeline_info(
                  pipeline_name,
 -                f"pipeline submitted: {public_api_endpoint}/#/runs/details/{run.id}",
 +                f"pipeline submitted: {public_api_endpoint}/{run.id}",
```
Comment (jstourac) [inline jupyter/datascience/ubi9-python-3.9/utils/processor_kfp.patch:25]: is this intentional?

**Comment by jstourac** [comment]
Thank you @harshad16 - this is awesome :slightly_smiling_face:

File: ci/check-json.sh (hunk 1 above)
Comment (harshad16) [reply ci/check-json.sh]: yeah true, refixed it. 
thanks :+1:

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -56,3 +57,17 @@ Pull requests are the best way to propose changes to the notebooks repository:
...
+
+### Linting
+
+To run linting tests, we use [pre-commit](https://pre-commit.com/).
+
+We have setup a [pre-commit](https://pre-commit.com) config file in [.pre-commit-config.yaml](.pre-commit-config.yaml).
+To [utilize pre-commit](https://pre-commit.com/#usage), install pre-commit with `pip3 install pre-commit` and then either:
+
+Run `pre-commit install` after you clone the repo, `pre-commit` will run automatically on git commit.
+   * If any one of the tests fail, add and commit the changes made by pre-commit. Once the pre-commit check passes, you can make your PR.
```
Comment (rkpattnaik780) [inline CONTRIBUTING.md:69]:
```suggestion
   * If any one of the tests fails, add and commit the changes made by pre-commit. Once the pre-commit check passes, you can make your PR.
```

**Review by rkpattnaik780 — COMMENTED** [review]
LGTM!

**Comment by harshad16** [comment]
Closing this for now,
will open this in future.

**Pull request closed without merging**